File: java/PalindromeLinkedList.java
/*
 * Filename: PalindromeLinkedList.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Mon 13 Jul 2015 12:32:07 PM EDT
 */

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
	public boolean isPalindrome(ListNode head) {
		if (head == null || head.next == null) {
			return true;
		}

		if (head.next.next == null) {
			return head.val == head.next.val;
		}

		ListNode p = head;
		ListNode p2 = head;

		while (p.next != null && p.next.next != null) {
			p = p.next.next;
			p2 = p2.next;
		}

		ListNode head2 = p2.next;
		p2.next = null;
		head = reverse(head);

		if (p.next == null) {
			return equals(head.next, head2);
		} else {
			return equals(head, head2);
		}
	}

	private ListNode reverse(ListNode head) {
		ListNode p = null;

		while (head != null) {
			ListNode tmp = head.next;
			head.next = p;
			p = head;
			head = tmp;
		}

		return p;
	}

	private boolean equals(ListNode head, ListNode head2) {
		while (head != null && head2 != null) {
			if (head.val != head2.val) {
				return false;
			}

			head = head.next;
			head2 = head2.next;
		}

		if (head != null || head2 != null) {
			return false;
		}

		return true;
	}
}

File: cpp/palindromeLinkedList.cpp
/*
 * Filename: palindromeLinkedList.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Tue 19 Jan 2016 07:35:41 PM EST
 * Description: 
 */

#include <iostream>

using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution
{
	public:
		bool isPalindrome(ListNode* head)
		{
			if (!head || !head->next)
			{
				return true;
			}

			if (!head->next->next)
			{
				return head->val == head->next->val;
			}

			ListNode* p = head;
			ListNode* q = head;

			while (p->next && p->next->next)
			{
				p = p->next->next;
				q = q->next;
			}

			ListNode* head2 = q->next;
			q->next = NULL;
			head = reverse(head);

			if (p->next)
			{
				return isEqual(head, head2);
			}
			else
			{
				return isEqual(head->next, head2);
			}
		}

	private:
		ListNode* reverse(ListNode* head)
		{
			ListNode* tail = NULL;
			ListNode* next; 
			while (head)
			{
				next = head->next;
				head->next = tail;
				tail = head;
				head = next;
			}

			return tail;
		}

		bool isEqual(ListNode* p, ListNode* q)
		{
			while (p && q)
			{
				if (p->val != q->val)
				{
					return false;
				}

				p = p->next;
				q = q->next;
			}

			return p == q;
		}
};
